Add an upgrading doc to help users switch to 6.0

sirosen opened the pull request:
The doc covers several issues, and explicitly is meant to address and resolve #485.

The doc is modeled on the marshmallow upgrading doc, and documents every backwards incompatible change which was made in 6.0, along with an example of impacted code and a rewrite which restores equivalent or near-equivalent behavior.

The doc mostly steers clear of trying to explain the decisions made, but it does provide some context, especially at the beginning when we have to show major signature changes in `use_args`/`use_kwargs`/`parse`.
Most of the suggested rewrites are simple, but a couple -- most notably the UnpackingDelimitedListSchema case -- are ideas which might be non-obvious to users.

Except when documenting specific changes to `parse` and related functions, everything is using `locations=("query",)` to specify "query params only" in the 5.x version of the code. That way, the translation to 6.x remains very simple in these cases.

@lafrech, I remember you saying when we did the 6.0 release that you were going to skip on writing an upgrade guide, since the changelog is pretty comprehensive. I agreed, but after seeing #483 and #485 , I figured I'd take a couple of hours and knock this out.

sirosen committed:
Add an upgrading doc to help users switch to 6.0

The doc covers several issues, and explicitly is meant to address and
resolve #485.

The doc is modeled on the marshmallow upgrading doc, and documents
every backwards incompatible change which was made in 6.0, along with
an example of impacted code and a rewrite which restores equivalent or
near-equivalent behavior.

The doc mostly steers clear of trying to explain the decisions made,
but it does provide some context, especially at the beginning when we
have to show major signature changes in
`use_args`/`use_kwargs`/`parse`.
Most of the suggested rewrites are simple, but a couple -- most
notably the UnpackingDelimitedListSchema case -- are ideas which might
be non-obvious to users.

Except when documenting specific changes to `parse` and related
functions, everything is using `locations=("query",)` to specify "query
params only" in the 5.x version of the code. That way, the translation
to 6.x remains very simple in these cases.

lafrech commented on docs/upgrading.rst:
```diff
@@ -0,0 +1,447 @@
+In webargs 6, the way that data is loaded from various locations and passed to
+your schemas changed significantly. This may require you to change several
+different parts of your webargs usage.
+
```
I tried to reword this, but in fact, I'd just remove it.

Next paragraphs speak for themselves.

sirosen replied:
👍 I'll remove it in a fixup commit.

lafrech commented on docs/upgrading.rst:
```diff
@@ -0,0 +1,447 @@
+your schemas changed significantly. This may require you to change several
+different parts of your webargs usage.
+
+Multiple locations are no longer supported in a single call
```
I think the convention is to capitalize titles.

sirosen replied:
Okay, I'll switch all of the titles to Title Case.

lafrech commented on docs/upgrading.rst:
```diff
@@ -0,0 +1,447 @@
+Multiple locations are no longer supported in a single call
+-----------------------------------------------------------
+
+And only JSON data is parsed by default.
```
The default location is JSON/body.

sirosen replied:
I take it this is a suggested rewording? I'll go with it. It's shorter than what I had.

lafrech commented on docs/upgrading.rst:
```diff
@@ -0,0 +1,447 @@
+
+In webargs 5.x, the schema you gave was used to pull data out of the request
+object. That data was compiled into a dictionary which was then passed to your
+schema.
```
In webargs 5.x, the deserialization schema was used to pull data out of the request
object. That data was compiled into a dictionary which was then passed to the schema.

lafrech commented on docs/upgrading.rst:
```diff
@@ -0,0 +1,447 @@
+One of the major changes in webargs 6.x allows the use of `unknown` parameter
+on schemas. This lets a schema decide what to do with fields not specified in
+the schema. In order to achieve this, webargs now passes the full data from
+the location you selected to your schema.
```
webargs now passes the full data from specified location to the schema.

lafrech commented on docs/upgrading.rst:
```diff
@@ -0,0 +1,447 @@
+    # webargs 6.x
+    @use_args({"q": ma.fields.String()}, location="query")
+    def viewfunc(request, q):
+        ...
```
I didn't double-check but I think request was always first and this change only affects multiple calls to `use_args`.

sirosen replied:
I'll check this and rewrite if I got it wrong.

lafrech reviewed (commented):
Thank you so much for this.

I made a few comments. Otherwise, good to go!

Also, I generally prefer to avoid addressing the reader directly, using "you", using "one" instead. Merely a matter of style and I don't really mind. This doc is too precious to block on this.

sirosen commented:
I like all of your rewording. I'll do all of these except for the pyramid one right now -- I want to check the behavior and make sure I have a correct example of the change.

sirosen committed:
Fixup the 6.0 upgrading doc

Mostly per review from @lafrech, but this also fixes the pyramid
parser change doc.
`request` is always passed first, but other argument ordering is
inverted in `@use_args`.

Additionally, this fixes all of the `use_args` calls to have less
confusing parameter names. Because it passes the args as a dict, use
the name `args` and show it can be accessed with `args.get("q")` in
each case. Naming the arg dict `q` after the only field used could
suggest to users that `use_args` passes fields individually in some
way.

sirosen commented:
I've just gotten to testing the PyramidParser behaviors and pushing the fix. I gave a new example in which there's a user-defined decorator passing additional data, so it clearly shows how `use_args` changes ordering.

I also noticed that the `use_args` calls in this doc were using a kind of confusing convention of naming the arg dict after the only field used, like
```python
@use_args({"q": fields.String()})
def foo(q):
    bar(q)
```

To make the behavior of `use_args` less confusing, I changed all of these to
```python
@use_args({"q": fields.String()})
def foo(args):
    bar(args.get("q"))
```

lafrech commented:
Sorry, I forgot about this.
